Check before casting Interceptors to CommonBean

benjamin-confino opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Resolves #31561

This code works around a customer issue in which we cast the FaultTolerence interceptor to CommonBean which promptly fails as it does not implement that method.

benjamin-confino committed:
Create a test for a fault tolerence annotation on an EJB

benjamin-confino committed:
Interceptors registered via a CDI extension will not be CommonBeans

Azquelt commented on dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/InterceptorChain.java:
```diff
@@ -129,15 +126,23 @@ private <S> Object invokeNextInterceptor() throws Exception {
-            S previous = (S) this.activeInterceptors.putIfAbsent(commonBean.getIdentifier(), interceptorInstance);
-            if (previous != null) {
-                interceptorInstance = previous;
+            if (interceptor instanceof CommonBean) {
```
```suggestion
            if (commonBean != null) {
```

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/retry/DummyBean.java:
```diff
@@ -0,0 +1,18 @@
+
+import javax.enterprise.context.RequestScoped;
+
+@RequestScoped
+public class DummyBean {
+
+}
```
This didn't change anything so it can be removed.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/retry/StatelessRetryServlet.java:
```diff
@@ -0,0 +1,39 @@
+    @EJB
+    private FaultTolerenceInterceptorOnEJB faultTolerenceInterceptorOnEJB;
+
+    @Test
+    //The fault tolerance CDI Extension does not fire events for methods on an EJB on these versions
+    @SkipForRepeat({ MicroProfileActions.MP13_ID, MicroProfileActions.MP20_ID })
+    public void testRetryOnStatelessBean(HttpServletRequest req, HttpServletResponse resp) throws Exception {
+        faultTolerenceInterceptorOnEJB.unstableMethod();
+        Assert.assertTrue(faultTolerenceInterceptorOnEJB.isPassed());
+    }
```
Currently this tests that fallback is working, it doesn't test retry at all (though it does test we don't hit the reported issue).

If we're trying to test retry on an EJB, we should just test retry.

benjamin-confino replied:
I'll rename this, testing the reported issue is the important thing for getting a fix to the customer

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/retry/FaultTolerenceInterceptorOnEJB.java:
```diff
@@ -0,0 +1,37 @@
+
+    private boolean passed = false;
+
+    @Retry(maxRetries = 1)
+    @Fallback(fallbackMethod = "fallbackMethod")
+    public void unstableMethod() throws TestException {
+        throw new TestException();
+    }
+
+    public void fallbackMethod() {
+        passed = true;
+    }
+
+    public boolean isPassed() {
+        return passed;
+    }
```
```suggestion
    private int executions = 0;

    @Retry(maxRetries = 1)
    @Fallback(fallbackMethod = "fallbackMethod")
    public void unstableMethod() throws TestException {
        throw new TestException();
    }

    public void fallbackMethod() {
        passed = true;
    }

    public boolean isPassed() {
        return passed;
    }

    public void resetExecutionCount() {
        executions = 0;
    }

    public int getExecutionCount() {
        return executions;
    }
```

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/retry/FaultToleranceOnEJBServlet.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    //The fault tolerance CDI Extension does not fire events for methods on an EJB on these versions
+    @SkipForRepeat({ MicroProfileActions.MP13_ID, MicroProfileActions.MP20_ID })
+    public void testRetryOnStatelessBean(HttpServletRequest req, HttpServletResponse resp) throws Exception {
+        faultTolerenceInterceptorOnEJB.unstableMethod();
+        Assert.assertTrue(faultTolerenceInterceptorOnEJB.isPassed());
+    }
```
```suggestion
    public void testRetryOnStatelessBean(HttpServletRequest req, HttpServletResponse resp) throws Exception {
        faultTolerenceInterceptorOnEJB.unstableMethod();
        Assert.assertTrue(faultTolerenceInterceptorOnEJB.isPassed());
        Assert.assertEquals(2, faultTolerenceInterceptorOnEJB.getExecutionCount());
    }
```

Azquelt reviewed (commented):
I think we should have basic testing for the FT annotations here.

I've suggested changes to test Retry but we should also cover the others.

Azquelt committed:
Add EJB CircuitBreaker test

Azquelt committed:
Add EJB bulkhead test

Azquelt committed:
Add FT EJB Async tests (broken)

benjamin-confino committed:
tests for Retry and timeout

benjamin-confino committed:
move stateless to common pacakge

benjamin-confino committed:
convert tests from a customer recreate into a full suite of tests for FT EJB integration

benjamin-confino committed:
cleanup

benjamin-confino committed:
rename server add concurrency feature for async test

Azquelt committed:
Consolidate common test code

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/FaultToleranceOnEJBTest.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Server(SERVER_NAME)
+    @TestServlets({
+                    @TestServlet(contextRoot = APP_NAME, servlet = FallbackOnEJBServlet.class),
+                    @TestServlet(contextRoot = APP_NAME, servlet = RetryOnEJBServlet.class),
+                    @TestServlet(contextRoot = APP_NAME, servlet = TimeOutOnEJBServlet.class),
+    })
```
Need to add the other test servlets.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/BarrierFactory.java:
```diff
@@ -0,0 +1,105 @@
+ */
+public class BarrierFactory implements AutoCloseable {
+
+    /** Timeout for awaiting barriers, should be longer than any test requires */
+    private static final int TIMEOUT = 10000;
+    /** Timeout for waiting for await() to be called on a barrier */
+    private static final int COMPLETION_TIMEOUT = 1000;
```
Need to check these timeout values vs. what we can depend on in our build systems.

Azquelt replied:
I've expanded the timeouts.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/ExecutionAssert.java:
```diff
@@ -0,0 +1,85 @@
+        public void run() throws Exception;
+    }
+
+    public static <E extends Exception> E assertThrows(Class<E> expected, ThrowingRunnable runnable) {
```
Should add Javadoc to the methods in this class.

benjamin-confino committed:
enable all test servlets

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=544d4de6-1c9c-43a2-b894-ad6cff5897fa) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino committed:
create (experimental) test for an custom interceptor on a retryed method

benjamin-confino committed:
Disable tests not yet passing and prepare others for delivery

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=37301874-fcc1-4a81-a92d-2efecb197675) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/async/FTAsyncEJBTestServlet.java:
```diff
@@ -0,0 +1,76 @@
+
+@SuppressWarnings("serial")
+@WebServlet("/ftasyncejb")
+@Mode(TestMode.EXPERIMENTAL)
+public class FTAsyncEJBTestServlet extends FATServlet {
```
Comment to say why this is disabled.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/bulkhead/BulkheadEJBTestServlet.java:
```diff
@@ -0,0 +1,83 @@
+
+@SuppressWarnings("serial")
+@WebServlet("/bulkheadejb")
+@Mode(TestMode.EXPERIMENTAL)
+public class BulkheadEJBTestServlet extends FATServlet {
```
Why's this one disabled?

benjamin-confino replied:
Caught exception attempting to call test method testBulkhead on servlet com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.bulkhead.BulkheadEJBTestServlet java.lang.AssertionError: Timeout waiting for future at com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.ExecutionAssert.assertThrows(ExecutionAssert.java:99) at com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.ExecutionAssert.assertThrowsEjbWrapped(ExecutionAssert.java:114) at com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.bulkhead.BulkheadEJBTestServlet.testBulkhead(BulkheadEJBTestServlet.java:62)

Azquelt replied:
We've changed it to skip the MP 2.0 repeat and added a comment to explain.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/circuitbreaker/CircuitBreakerEJBTestServlet.java:
```diff
@@ -0,0 +1,71 @@
+    @SkipForRepeat(MicroProfileActions.MP20_ID)
+    @Test
+    @ExpectedFFDC("org.eclipse.microprofile.faulttolerance.exceptions.CircuitBreakerOpenException")
+    public void testEjbCircuitBreaker() {
```
Based on our discussion, I thought that the problem was that line 53 was throwing `TestException` rather than `CircuitBreakerOpenException`?

I don't mind leaving it with the test disabled for this repeat, but let's not leave misleading comments for future us.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/fallback/FallbackOnEJBServlet.java:
```diff
@@ -0,0 +1,39 @@
+    @Test
+    //The fault tolerance CDI Extension does not fire events for methods on an EJB on these versions
+    //@SkipForRepeat({ MicroProfileActions.MP13_ID, MicroProfileActions.MP20_ID })
+    public void testFallbackOnEJB(HttpServletRequest req, HttpServletResponse resp) throws Exception {
```
What's this commented out skip for?

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/interceptedretry/InterceptedRetryOnEJB.java:
```diff
@@ -0,0 +1,39 @@
+    //Test that the LoggingInterceptor is called every retry
+    @Retry(maxRetries = 10)
+    @LogInterceptorBinding
+    public int retryEventuallyPass() throws TestException {
```
```suggestion
    public static final int MAX = 5;
    public static final int MAX_PLUS_FENCEPOST = 6; //The intercepter will be called for
    private static int counter = 0;

    //Test that the LoggingInterceptor is called every retry
    @Retry(maxRetries = MAX)
    @LogInterceptorBinding
    public int retryEventuallyPass() throws TestException {
```

I think it should be this, based on the rest of the test.

Given that the test is disabled because the functionality is not working, it's hard to be sure.

benjamin-confino replied:
It shouldn't, that's the maximum number of retries, which has to be higher than the number of times the method fails. I'll make this clearer.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/retry/RetryOnEJBServlet.java:
```diff
@@ -0,0 +1,59 @@
+        //This method will fail until it has failed the Maximum number of times it was configured to
+        //do so. Fault Tolerence will make it retry each time.
+
+        //When it has reached the maximum it will successfully return the number of times it has run.
+        int result = ejb.retryEventuallyPass();
+        Assert.assertEquals(RetryOnEJB.MAX, result);
```
I find these tests hard to read. It's all very well having constants, but it's not clear what the constants mean.

Maybe in future we'll come around again and find a clearer way of writing the test so it's easy to see what retry scenarios we're testing.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/timeout/ResultsRecord.java:
```diff
@@ -0,0 +1,8 @@
+package com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.timeout;
+
+public class ResultsRecord {
+
+    public boolean testMethodCalled = false;
+    public boolean testMethodRecievedInteruptException = false;
+    public boolean testMethodContinuedPastInterruptException = false;
+}
```
Missing copyright header

Azquelt replied:
Still missing copyright header. Unfortunate, we removed the class and so didn't need a header, but then we put it back again.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/timeout/TimeOutOnEJBServlet.java:
```diff
@@ -0,0 +1,58 @@
+    @Test
+    //The fault tolerance CDI Extension does not fire events for methods on an EJB on these versions
+    //@SkipForRepeat({ MicroProfileActions.MP13_ID, MicroProfileActions.MP20_ID })
+    public void testTimeoutOnEJB(HttpServletRequest req, HttpServletResponse resp) throws Exception {
```
Commented out restriction again.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/timeout/TimeOutOnEJBServlet.java:
```diff
@@ -0,0 +1,58 @@
+            Assert.assertTrue(record.testMethodRecievedInteruptException);
+            Assert.assertFalse(record.testMethodContinuedPastInterruptException);
+        }
+        Assert.assertTrue(caughtException);
```
```suggestion
        assertThrowsEjbWrapped(TimeoutException.class, () -> ejb.testMethodThatTimesOut(record));
```

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/timeout/TimeOutOnEJBServlet.java:
```diff
@@ -0,0 +1,58 @@
+    //@SkipForRepeat({ MicroProfileActions.MP13_ID, MicroProfileActions.MP20_ID })
+    public void testNoTimeoutOnEJB(HttpServletRequest req, HttpServletResponse resp) throws Exception {
+        Assert.assertEquals(TimeoutOnEJB.SOME_VALUE, ejb.testMethodThatWontTimeOut());
+    }
```
Commented out restriction

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/timeout/TimeOutOnEJBServlet.java:
```diff
@@ -0,0 +1,58 @@
+    //The fault tolerance CDI Extension does not fire events for methods on an EJB on these versions
+    //@SkipForRepeat({ MicroProfileActions.MP13_ID, MicroProfileActions.MP20_ID })
+    public void testNoTimeoutOnEJB(HttpServletRequest req, HttpServletResponse resp) throws Exception {
+        Assert.assertEquals(TimeoutOnEJB.SOME_VALUE, ejb.testMethodThatWontTimeOut());
```
```suggestion
        assertReturns(() -> ejb.testMethodThatWontTimeOut());
```

The way the test is currently written makes it look like the return value is important, but it isn't really.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/publish/servers/FTAnnotationsDirectlyOnEjb/server.xml:
```diff
@@ -0,0 +1,28 @@
+   
+    Contributors:
+        IBM Corporation - initial API and implementation
+ -->
```
```suggestion
    SPDX-License-Identifier: EPL-2.0
 -->
```

Azquelt commented on dev/com.ibm.ws.cdi.ejb_fat/fat/src/com/ibm/ws/cdi/ejb/tests/SerializeEjbTest.java:
```diff
@@ -0,0 +1,209 @@
+ */
+@Mode(TestMode.EXPERIMENTAL)
+@RunWith(FATRunner.class)
+public class SerializeEjbTest {
```
I'm unsure what we should do with this test.

On the one hand, it currently doesn't work and we do want to make it work (or at least find out why it doesn't work). Having it in the codebase makes it easier to work on.

On the other hand, it's currently in an undetermined state. We know it doesn't pass but we're not sure why or whether it should pass. Having it in the code base in this state is confusing, it's not clear what the test is testing, how it works or whether it should pass.

I'm leaning towards taking it out of this PR, putting it in its own branch and raising a new issue to get it integrated. WDYT?

Azquelt replied:
We took it out

Azquelt committed:
Add comments to ExecutionAssert

benjamin-confino committed:
code review cleanup

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/timeout/TimeOutOnEJBServlet.java:
```diff
@@ -31,28 +32,13 @@ public class TimeOutOnEJBServlet extends FATServlet {
-            Assert.assertTrue(e.getCausedByException() instanceof TimeoutException);
-            Assert.assertTrue(record.testMethodCalled);
-            Assert.assertTrue(record.testMethodRecievedInteruptException);
-            Assert.assertFalse(record.testMethodContinuedPastInterruptException);
```
Sorry, I was wrong when I said to remove these checks - eyes just skipped over them since they were within the catch block which was replaced with `assertThrowsEjbWrapped`.

I can restore the checks.

Azquelt committed:
Enable Bulkhead test

Accidentally left in experimental because it doesn't work on MP 2.0.
Added comment to explain.

Azquelt committed:
Explain why EJB + FT async test is disabled

Azquelt committed:
Add comments and remove warnings

Azquelt committed:
Enable EJB + Fallback test

Azquelt committed:
Extend FT + EJB timeout values

Build hardware can be slow, increase these so they're more in line with
other tests.

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e6f34f9c-de08-4bbd-b587-26faf58b973f) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=55c61795-5bbd-4ea2-a5a5-42ba77bc6fe5) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/publish/servers/FTAnnotationsDirectlyOnEjb/server.xml:
```diff
@@ -0,0 +1,25 @@
+
+    <include location="../fatTestPorts.xml" />
+
+<logging traceSpecification="**=info:EJBContainer=all" traceFileName="logs/trace.log"/>
```
I suspect we don't need this trace.

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/interceptedretry/LogInterceptorBinding.java:
```diff
@@ -0,0 +1,18 @@
+package com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.interceptedretry;
```
Missing copyright header

Azquelt commented on dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/tests/stateless/interceptedretry/LoggingInterceptor.java:
```diff
@@ -0,0 +1,18 @@
+package com.ibm.websphere.microprofile.faulttolerance_fat.tests.stateless.interceptedretry;
```
Missing copyright header

Azquelt reviewed (changes requested):
Three missing copyright headers, otherwise this looks fine.

benjamin-confino committed:
copyright headers and remove tracce statement

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=1d2e28a7-7ce5-4764-8fb8-d0dafa049a72) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._
